application/frontend/controllers/MfaController.php

```diff
@@ -0,0 +1,138 @@
+<?php
+namespace frontend\controllers;
+
+use common\components\MfaBackendInterface;
+use common\models\Mfa;
+use common\models\User;
+use frontend\components\BaseRestController;
+use yii\web\BadRequestHttpException;
+use yii\web\NotFoundHttpException;
+use yii\web\ServerErrorHttpException;
+
+class MfaController extends BaseRestController
+{
+
+    /**
+     * Create new MFA record
+     * @return array
+     * @throws BadRequestHttpException
+     */
+    public function actionCreate()
+    {
+        $req = \Yii::$app->request;
+        $type = $req->getBodyParam('type','invalid');
+        if ( ! Mfa::isValidType($type)) {
+            throw new BadRequestHttpException('The provided type is not a supported MFA type', 1506695647);
+        }
+
+        $employeeId = $req->getBodyParams('employee_id');
+        if ( is_null($employeeId)) {
+            throw new BadRequestHttpException('employee_id is required', 1506695722);
+        }
+
+        $user = User::findOne(['employee_id' => $employeeId]);
+        if ($user === null) {
+            throw new BadRequestHttpException('User not found', 1506695733);
+        }
+
+        $label = $req->getBodyParam('label');
+
+        return Mfa::create($user->id, $type, $label);
+    }
+
+    /**
+     * Verify value with MFA backend
+     * @param int $id
+     * @throws BadRequestHttpException
+     * @throws NotFoundHttpException
+     * @return User
+     */
+    public function actionVerify(int $id)
+    {
+        $req = \Yii::$app->request;
+        $value = $req->getBodyParam('value');
+        if ($value === null) {
+            throw new BadRequestHttpException('Value is required for verification', 1506697678);
+        }
+
+        $employeeId = $req->getBodyParam('employee_id');
+        if ($employeeId == null) {
+            throw new BadRequestHttpException("employee_id is required");
+        }
+
+        $user = User::findOne(['employee_id' => $employeeId]);
+        if ($user == null) {
+            throw new BadRequestHttpException("Invalid employee_id");
+        }
+
+        $mfa = Mfa::findOne(['id' => $id, 'user_id' => $user->id]);
+        if ($mfa === null) {
+            throw new NotFoundHttpException(
+                sprintf('MFA record for id %s not found', $id),
+                1506697604
+            );
+        }
+
+        // Strip spaces from $value if string
+        if (is_string($value)) {
+            $value = str_replace(' ', '', $value);
```

Comment:
Is it possible for a U2F value to contain a necessary space?

Reply:
for u2f it comes in as an array, parsed out by yii's json parser, so the value is not a string and doesn't get caught by this